### PR TITLE
Correctly format filter ws requests

### PR DESCRIPTION
**Pull request by mrwillis**
# Description

Currently the websocket subscriptions fail because the unmarshaller unmarshals JSON numbers into float64 and not integer, which is what the formatting function expects, resulting in stuff like this

```
CONNECTED

SENT: {"id": 1, "method": "eth_subscribe", "params": ["newHeads", {}]}

RECEIVED: {"jsonrpc":"2.0","id":%!d(float64=1),"result":"4e31841a-c61e-442a-b1a8-de998db7f164"}
```

Moreover, it does not correctly consider strings, which are also valid under the spec. 

https://www.jsonrpc.org/specification

This results in websocket requests over JSON-RPC to timeout because it never sees a response with the same `id` sent in.

See this as to the unmarshalling behaviour

https://stackoverflow.com/questions/55436628/json-decoded-value-is-treated-as-float64-instead-of-int/55436758. 

I will also add error checking so that the `id` field is the correct format. 

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [x] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

File: jsonrpc/dispatcher_test.go [hunk 1]
```diff
@@ -25,6 +25,26 @@ func expectJSONResult(data []byte, v interface{}) error {
 	return nil
 }
 
+func TestDispatcherWebsocketStringId(t *testing.T) {
```
Comment (Kourin1996) [inline jsonrpc/dispatcher_test.go]: This and `TestDispatcherWebsocketInvalidRequestFormats` are similar to each other and the targets are same. Can we put this 2 cases into one test by using subtests?

**Review by zivkovicmilos — APPROVED** [review]
Looks good to me 💯

File: jsonrpc/dispatcher.go [hunk 2]
```diff
@@ -113,6 +114,26 @@ type wsConn interface {
...
+	case string:
+		return fmt.Sprintf(`{"jsonrpc":"2.0","id":%s,"result":"%s"}`, t, resp), nil
+	case float64:
+		if t == math.Trunc(t) {
+			return fmt.Sprintf(`{"jsonrpc":"2.0","id":%d,"result":"%s"}`, int(t), resp), nil
+		} else {
+			return "", invalidJSONRequest
+		}
+	case nil:
+		return fmt.Sprintf(`{"jsonrpc":"2.0","result":"%s"}`, resp), nil
```
Comment (Vuksan) [inline jsonrpc/dispatcher.go]: I think that by the [JSON-RPC spec](https://www.jsonrpc.org/specification) `id` is required in the `response object`, at least as `Nil`:
Screen Shot 2021-08-05 at 16 16 21 PM

**Review by Vuksan — COMMENTED** [review]
Thank you for noticing this and adding the fix.
I left a single comment, just take another look at the spec

File: jsonrpc/dispatcher.go (hunk 2 above)
Comment (mrwillis) [reply jsonrpc/dispatcher.go]: Nice catch will fix

File: jsonrpc/dispatcher_test.go (hunk 1 above)
Comment (mrwillis) [reply jsonrpc/dispatcher_test.go]: Done

File: jsonrpc/dispatcher.go (hunk 2 above)
Comment (mrwillis) [reply jsonrpc/dispatcher.go]: Okay added that

**Review by Kourin1996 — APPROVED** [review]
LGTM 👏 Thank you for reporting the problem and fixing.

**Review by Vuksan — APPROVED** [review]
🙌

**Pull request merged**